Reduce number of queries required for content tagging endpoints

pomegranited opened the pull request:
## Description

Tweak the queries used in the content tagging views to reduce the number of database queries made, and improve response time.

Content tagging is disabled by default, so no one should be affected by this change.

## Supporting information

Closes https://github.com/openedx/modular-learning/issues/185

Depends on https://github.com/openedx/openedx-learning/pull/157
Relates to https://github.com/open-craft/edx-platform/pull/633

Internal-ref: [FAL-3646](https://tasks.opencraft.com/browse/FAL-3646)

## Testing instructions

These changes should be validated by the unit tests.
Please also review the sql queries shown below to ensure that no duplicate queries are being performed, and there's no other obvious ways to reduce these query counts.

### List Taxonomies

<details><summary>11 queries for Taxonomy admins (reduced from 16)</summary>

```
E   Captured queries were:
E   1. SELECT "django_site"."id", "django_site"."domain", "django_site"."name" FROM "django_site" WHERE "django_site"."domain" LIKE 'testserver' ESCAPE '\' LIMIT 21
E   2. SELECT "student_userstanding"."id", "student_userstanding"."user_id", "student_userstanding"."account_status", "student_userstanding"."changed_by_id", "student_userstanding"."standing_last_changed_at" FROM "student_userstanding" WHERE "student_userstanding"."user_id" IS NULL LIMIT 21
E   3. SELECT "theming_sitetheme"."id", "theming_sitetheme"."site_id", "theming_sitetheme"."theme_dir_name" FROM "theming_sitetheme" WHERE "theming_sitetheme"."site_id" = 1 ORDER BY "theming_sitetheme"."id" ASC LIMIT 1
E   4. SAVEPOINT "s140265752491840_x3439"
E   5. SELECT COUNT(*) FROM (SELECT "oel_tagging_taxonomy"."id" AS "col1" FROM "oel_tagging_taxonomy" LEFT OUTER JOIN "oel_tagging_tag" ON ("oel_tagging_taxonomy"."id" = "oel_tagging_tag"."taxonomy_id") WHERE ("oel_tagging_taxonomy"."enabled" AND EXISTS(SELECT 1 AS "a" FROM "content_tagging_taxonomyorg" U0 LEFT OUTER JOIN "organizations_organization" U2 ON (U0."org_id" = U2."id") WHERE (U0."rel_type" = 'OWN' AND U0."taxonomy_id" = ("oel_tagging_taxonomy"."id") AND (U0."org_id" IS NULL OR U2."short_name" = 'orgA')) LIMIT 1)) GROUP BY 1) subquery
E   6. SELECT "oel_tagging_taxonomy"."id", "oel_tagging_taxonomy"."name", "oel_tagging_taxonomy"."description", "oel_tagging_taxonomy"."enabled", "oel_tagging_taxonomy"."allow_multiple", "oel_tagging_taxonomy"."allow_free_text", "oel_tagging_taxonomy"."visible_to_authors", "oel_tagging_taxonomy"."export_id", "oel_tagging_taxonomy"."_taxonomy_class", COUNT(DISTINCT "oel_tagging_tag"."id") AS "tags_count" FROM "oel_tagging_taxonomy" LEFT OUTER JOIN "oel_tagging_tag" ON ("oel_tagging_taxonomy"."id" = "oel_tagging_tag"."taxonomy_id") WHERE ("oel_tagging_taxonomy"."enabled" AND EXISTS(SELECT 1 AS "a" FROM "content_tagging_taxonomyorg" U0 LEFT OUTER JOIN "organizations_organization" U2 ON (U0."org_id" = U2."id") WHERE (U0."rel_type" = 'OWN' AND U0."taxonomy_id" = ("oel_tagging_taxonomy"."id") AND (U0."org_id" IS NULL OR U2."short_name" = 'orgA')) LIMIT 1)) GROUP BY "oel_tagging_taxonomy"."id", "oel_tagging_taxonomy"."name", "oel_tagging_taxonomy"."description", "oel_tagging_taxonomy"."enabled", "oel_tagging_taxonomy"."allow_multiple", "oel_tagging_taxonomy"."allow_free_text", "oel_tagging_taxonomy"."visible_to_authors", "oel_tagging_taxonomy"."export_id", "oel_tagging_taxonomy"."_taxonomy_class" ORDER BY "oel_tagging_taxonomy"."name" ASC, "oel_tagging_taxonomy"."id" ASC LIMIT 4
E   7. SELECT "content_tagging_taxonomyorg"."id", "content_tagging_taxonomyorg"."taxonomy_id", "content_tagging_taxonomyorg"."org_id", "content_tagging_taxonomyorg"."rel_type" FROM "content_tagging_taxonomyorg" WHERE "content_tagging_taxonomyorg"."taxonomy_id" IN (3, 5, 7, 11)
E   8. SELECT "organizations_organization"."id", "organizations_organization"."created", "organizations_organization"."modified", "organizations_organization"."name", "organizations_organization"."short_name", "organizations_organization"."description", "organizations_organization"."logo", "organizations_organization"."active" FROM "organizations_organization" WHERE "organizations_organization"."id" IN (2, 1)
E   9. SELECT COUNT(*) AS "__count" FROM "oel_tagging_tag" WHERE "oel_tagging_tag"."taxonomy_id" = 3
E   10. RELEASE SAVEPOINT "s140265752491840_x3439"
E   11. SELECT "user_api_userpreference"."id", "user_api_userpreference"."user_id", "user_api_userpreference"."key", "user_api_userpreference"."value" FROM "user_api_userpreference" WHERE ("user_api_userpreference"."key" = 'pref-lang' AND "user_api_userpreference"."user_id" = 2) LIMIT 21
```
</details>

<details><summary>16 queries for content authors (unchanged)</summary>

```
E   Captured queries were:
E   1. SELECT "django_site"."id", "django_site"."domain", "django_site"."name" FROM "django_site" WHERE "django_site"."domain" LIKE 'testserver' ESCAPE '\' LIMIT 21
E   2. SELECT "student_userstanding"."id", "student_userstanding"."user_id", "student_userstanding"."account_status", "student_userstanding"."changed_by_id", "student_userstanding"."standing_last_changed_at" FROM "student_userstanding" WHERE "student_userstanding"."user_id" IS NULL LIMIT 21
E   3. SELECT "theming_sitetheme"."id", "theming_sitetheme"."site_id", "theming_sitetheme"."theme_dir_name" FROM "theming_sitetheme" WHERE "theming_sitetheme"."site_id" = 1 ORDER BY "theming_sitetheme"."id" ASC LIMIT 1
E   4. SAVEPOINT "s140700559607616_x4616"
E   5. SELECT "organizations_organization"."id", "organizations_organization"."created", "organizations_organization"."modified", "organizations_organization"."name", "organizations_organization"."short_name", "organizations_organization"."description", "organizations_organization"."logo", "organizations_organization"."active" FROM "organizations_organization"
E   6. SELECT "student_courseaccessrole"."id", "student_courseaccessrole"."user_id", "student_courseaccessrole"."org", "student_courseaccessrole"."course_id", "student_courseaccessrole"."role" FROM "student_courseaccessrole" WHERE "student_courseaccessrole"."user_id" = 6
E   7. SELECT "content_libraries_contentlibrary"."id", "content_libraries_contentlibrary"."org_id", "organizations_organization"."id", "organizations_organization"."created", "organizations_organization"."modified", "organizations_organization"."name", "organizations_organization"."short_name", "organizations_organization"."description", "organizations_organization"."logo", "organizations_organization"."active" FROM "content_libraries_contentlibrary" LEFT OUTER JOIN "content_libraries_contentlibrarypermission" ON ("content_libraries_contentlibrary"."id" = "content_libraries_contentlibrarypermission"."library_id") INNER JOIN "organizations_organization" ON ("content_libraries_contentlibrary"."org_id" = "organizations_organization"."id") WHERE ("content_libraries_contentlibrary"."allow_public_read" OR "content_libraries_contentlibrarypermission"."user_id" = 6 OR "content_libraries_contentlibrarypermission"."group_id" IN (SELECT U0."id" FROM "auth_group" U0 INNER JOIN "auth_user_groups" U1 ON (U0."id" = U1."group_id") WHERE U1."user_id" = 6))
E   8. SELECT COUNT(*) FROM (SELECT "oel_tagging_taxonomy"."id" AS "col1" FROM "oel_tagging_taxonomy" LEFT OUTER JOIN "oel_tagging_tag" ON ("oel_tagging_taxonomy"."id" = "oel_tagging_tag"."taxonomy_id") WHERE ("oel_tagging_taxonomy"."enabled" AND EXISTS(SELECT 1 AS "a" FROM "content_tagging_taxonomyorg" U0 LEFT OUTER JOIN "organizations_organization" U2 ON (U0."org_id" = U2."id") WHERE (U0."rel_type" = 'OWN' AND U0."taxonomy_id" = ("oel_tagging_taxonomy"."id") AND (U0."org_id" IS NULL OR U2."short_name" = 'orgA')) LIMIT 1) AND (EXISTS(SELECT 1 AS "a" FROM "content_tagging_taxonomyorg" U0 WHERE (U0."rel_type" = 'OWN' AND U0."taxonomy_id" = ("oel_tagging_taxonomy"."id") AND (U0."org_id" IS NULL OR U0."org_id" IN (1))) LIMIT 1) AND "oel_tagging_taxonomy"."enabled")) GROUP BY 1) subquery
E   9. SELECT "oel_tagging_taxonomy"."id", "oel_tagging_taxonomy"."name", "oel_tagging_taxonomy"."description", "oel_tagging_taxonomy"."enabled", "oel_tagging_taxonomy"."allow_multiple", "oel_tagging_taxonomy"."allow_free_text", "oel_tagging_taxonomy"."visible_to_authors", "oel_tagging_taxonomy"."export_id", "oel_tagging_taxonomy"."_taxonomy_class", COUNT(DISTINCT "oel_tagging_tag"."id") AS "tags_count" FROM "oel_tagging_taxonomy" LEFT OUTER JOIN "oel_tagging_tag" ON ("oel_tagging_taxonomy"."id" = "oel_tagging_tag"."taxonomy_id") WHERE ("oel_tagging_taxonomy"."enabled" AND EXISTS(SELECT 1 AS "a" FROM "content_tagging_taxonomyorg" U0 LEFT OUTER JOIN "organizations_organization" U2 ON (U0."org_id" = U2."id") WHERE (U0."rel_type" = 'OWN' AND U0."taxonomy_id" = ("oel_tagging_taxonomy"."id") AND (U0."org_id" IS NULL OR U2."short_name" = 'orgA')) LIMIT 1) AND (EXISTS(SELECT 1 AS "a" FROM "content_tagging_taxonomyorg" U0 WHERE (U0."rel_type" = 'OWN' AND U0."taxonomy_id" = ("oel_tagging_taxonomy"."id") AND (U0."org_id" IS NULL OR U0."org_id" IN (1))) LIMIT 1) AND "oel_tagging_taxonomy"."enabled")) GROUP BY "oel_tagging_taxonomy"."id", "oel_tagging_taxonomy"."name", "oel_tagging_taxonomy"."description", "oel_tagging_taxonomy"."enabled", "oel_tagging_taxonomy"."allow_multiple", "oel_tagging_taxonomy"."allow_free_text", "oel_tagging_taxonomy"."visible_to_authors", "oel_tagging_taxonomy"."export_id", "oel_tagging_taxonomy"."_taxonomy_class" ORDER BY "oel_tagging_taxonomy"."name" ASC, "oel_tagging_taxonomy"."id" ASC LIMIT 4
E   10. SELECT "content_tagging_taxonomyorg"."id", "content_tagging_taxonomyorg"."taxonomy_id", "content_tagging_taxonomyorg"."org_id", "content_tagging_taxonomyorg"."rel_type" FROM "content_tagging_taxonomyorg" WHERE "content_tagging_taxonomyorg"."taxonomy_id" IN (3, 5, 7, 11)
E   11. SELECT "organizations_organization"."id", "organizations_organization"."created", "organizations_organization"."modified", "organizations_organization"."name", "organizations_organization"."short_name", "organizations_organization"."description", "organizations_organization"."logo", "organizations_organization"."active" FROM "organizations_organization" WHERE "organizations_organization"."id" IN (2, 1)
E   12. SELECT COUNT(*) AS "__count" FROM "oel_tagging_tag" WHERE "oel_tagging_tag"."taxonomy_id" = 3
E   13. SELECT "django_content_type"."app_label", "auth_permission"."codename" FROM "auth_permission" INNER JOIN "auth_user_user_permissions" ON ("auth_permission"."id" = "auth_user_user_permissions"."permission_id") INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE "auth_user_user_permissions"."user_id" = 6
E   14. SELECT "django_content_type"."app_label", "auth_permission"."codename" FROM "auth_permission" INNER JOIN "auth_group_permissions" ON ("auth_permission"."id" = "auth_group_permissions"."permission_id") INNER JOIN "auth_group" ON ("auth_group_permissions"."group_id" = "auth_group"."id") INNER JOIN "auth_user_groups" ON ("auth_group"."id" = "auth_user_groups"."group_id") INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE "auth_user_groups"."user_id" = 6
E   15. RELEASE SAVEPOINT "s140700559607616_x4616"
E   16. SELECT "user_api_userpreference"."id", "user_api_userpreference"."user_id", "user_api_userpreference"."key", "user_api_userpreference"."value" FROM "user_api_userpreference" WHERE ("user_api_userpreference"."key" = 'pref-lang' AND "user_api_userpreference"."user_id" = 6) LIMIT 21
```
</details>

### Get object tags

*Note:* Queries shown include changes from https://github.com/open-craft/edx-platform/pull/633.

<details><summary>8 queries for taxonomy admins (increased from 7, added cross-org check)</summary>

```
E   Captured queries were:
E   1. SELECT "django_site"."id", "django_site"."domain", "django_site"."name" FROM "django_site" WHERE "django_site"."domain" LIKE 'testserver' ESCAPE '\' LIMIT 21
E   2. SELECT "student_userstanding"."id", "student_userstanding"."user_id", "student_userstanding"."account_status", "student_userstanding"."changed_by_id", "student_userstanding"."standing_last_changed_at" FROM "student_userstanding" WHERE "student_userstanding"."user_id" IS NULL LIMIT 21
E   3. SELECT "theming_sitetheme"."id", "theming_sitetheme"."site_id", "theming_sitetheme"."theme_dir_name" FROM "theming_sitetheme" WHERE "theming_sitetheme"."site_id" = 1 ORDER BY "theming_sitetheme"."id" ASC LIMIT 1
E   4. SAVEPOINT "s140265752491840_x1427"
E   5. SELECT "oel_tagging_objecttag"."id", "oel_tagging_objecttag"."object_id", "oel_tagging_objecttag"."taxonomy_id", "oel_tagging_objecttag"."tag_id", "oel_tagging_objecttag"."_name", "oel_tagging_objecttag"."_value", LOWER(COALESCE(T6."value" || '        ', '') || COALESCE(COALESCE(T5."value" || '     ', '') || COALESCE(COALESCE(T4."value" || '        ', '') || COALESCE(COALESCE(COALESCE("oel_tagging_tag"."value", "oel_tagging_objecttag"."_value"), '') || COALESCE('       ', ''), ''), ''), '')) AS "sort_key", COALESCE("oel_tagging_taxonomy"."name", "oel_tagging_objecttag"."_name") AS "taxonomy_name", "oel_tagging_taxonomy"."id", "oel_tagging_taxonomy"."name", "oel_tagging_taxonomy"."description", "oel_tagging_taxonomy"."enabled", "oel_tagging_taxonomy"."allow_multiple", "oel_tagging_taxonomy"."allow_free_text", "oel_tagging_taxonomy"."visible_to_authors", "oel_tagging_taxonomy"."export_id", "oel_tagging_taxonomy"."_taxonomy_class", "oel_tagging_tag"."id", "oel_tagging_tag"."taxonomy_id", "oel_tagging_tag"."parent_id", "oel_tagging_tag"."value", "oel_tagging_tag"."external_id", T4."id", T4."taxonomy_id", T4."parent_id", T4."value", T4."external_id", T5."id", T5."taxonomy_id", T5."parent_id", T5."value", T5."external_id" FROM "oel_tagging_objecttag" INNER JOIN "oel_tagging_taxonomy" ON ("oel_tagging_objecttag"."taxonomy_id" = "oel_tagging_taxonomy"."id") LEFT OUTER JOIN "oel_tagging_tag" ON ("oel_tagging_objecttag"."tag_id" = "oel_tagging_tag"."id") LEFT OUTER JOIN "oel_tagging_tag" T4 ON ("oel_tagging_tag"."parent_id" = T4."id") LEFT OUTER JOIN "oel_tagging_tag" T5 ON (T4."parent_id" = T5."id") LEFT OUTER JOIN "oel_tagging_tag" T6 ON (T5."parent_id" = T6."id") WHERE ("oel_tagging_objecttag"."object_id" = 'course-v1:orgA+101+test' AND NOT (NOT "oel_tagging_taxonomy"."enabled" AND "oel_tagging_taxonomy"."enabled" IS NOT NULL) AND NOT ("oel_tagging_objecttag"."taxonomy_id" IS NULL) AND NOT ("oel_tagging_objecttag"."tag_id" IS NULL AND NOT "oel_tagging_taxonomy"."allow_free_text")) ORDER BY 8 ASC, 7 ASC
E   6. SELECT "content_tagging_taxonomyorg"."id", "content_tagging_taxonomyorg"."taxonomy_id", "content_tagging_taxonomyorg"."org_id", "content_tagging_taxonomyorg"."rel_type" FROM "content_tagging_taxonomyorg" WHERE "content_tagging_taxonomyorg"."taxonomy_id" = 5
E   7. RELEASE SAVEPOINT "s140265752491840_x1427"
E   8. SELECT "user_api_userpreference"."id", "user_api_userpreference"."user_id", "user_api_userpreference"."key", "user_api_userpreference"."value" FROM "user_api_userpreference" WHERE ("user_api_userpreference"."key" = 'pref-lang' AND "user_api_userpreference"."user_id" = 2) LIMIT 21
```
</details>

<details><summary>11 queries for content authors (reduced from 19)</summary>

```
E   Captured queries were:
E   1. SELECT "django_site"."id", "django_site"."domain", "django_site"."name" FROM "django_site" WHERE "django_site"."domain" LIKE 'testserver' ESCAPE '\' LIMIT 21
E   2. SELECT "student_userstanding"."id", "student_userstanding"."user_id", "student_userstanding"."account_status", "student_userstanding"."changed_by_id", "student_userstanding"."standing_last_changed_at" FROM "student_userstanding" WHERE "student_userstanding"."user_id" IS NULL LIMIT 21
E   3. SELECT "theming_sitetheme"."id", "theming_sitetheme"."site_id", "theming_sitetheme"."theme_dir_name" FROM "theming_sitetheme" WHERE "theming_sitetheme"."site_id" = 1 ORDER BY "theming_sitetheme"."id" ASC LIMIT 1
E   4. SAVEPOINT "s140700559607616_x2209"
E   5. SELECT "student_courseaccessrole"."id", "student_courseaccessrole"."user_id", "student_courseaccessrole"."org", "student_courseaccessrole"."course_id", "student_courseaccessrole"."role" FROM "student_courseaccessrole" WHERE "student_courseaccessrole"."user_id" = 6
E   6. SELECT "organizations_organization"."id", "organizations_organization"."created", "organizations_organization"."modified", "organizations_organization"."name", "organizations_organization"."short_name", "organizations_organization"."description", "organizations_organization"."logo", "organizations_organization"."active" FROM "organizations_organization"
E   7. SELECT "content_libraries_contentlibrary"."id", "content_libraries_contentlibrary"."org_id", "organizations_organization"."id", "organizations_organization"."created", "organizations_organization"."modified", "organizations_organization"."name", "organizations_organization"."short_name", "organizations_organization"."description", "organizations_organization"."logo", "organizations_organization"."active" FROM "content_libraries_contentlibrary" LEFT OUTER JOIN "content_libraries_contentlibrarypermission" ON ("content_libraries_contentlibrary"."id" = "content_libraries_contentlibrarypermission"."library_id") INNER JOIN "organizations_organization" ON ("content_libraries_contentlibrary"."org_id" = "organizations_organization"."id") WHERE ("content_libraries_contentlibrary"."allow_public_read" OR "content_libraries_contentlibrarypermission"."user_id" = 6 OR "content_libraries_contentlibrarypermission"."group_id" IN (SELECT U0."id" FROM "auth_group" U0 INNER JOIN "auth_user_groups" U1 ON (U0."id" = U1."group_id") WHERE U1."user_id" = 6))
E   8. SELECT "oel_tagging_objecttag"."id", "oel_tagging_objecttag"."object_id", "oel_tagging_objecttag"."taxonomy_id", "oel_tagging_objecttag"."tag_id", "oel_tagging_objecttag"."_name", "oel_tagging_objecttag"."_value", LOWER(COALESCE(T6."value" || '        ', '') || COALESCE(COALESCE(T5."value" || '     ', '') || COALESCE(COALESCE(T4."value" || '        ', '') || COALESCE(COALESCE(COALESCE("oel_tagging_tag"."value", "oel_tagging_objecttag"."_value"), '') || COALESCE('       ', ''), ''), ''), '')) AS "sort_key", COALESCE("oel_tagging_taxonomy"."name", "oel_tagging_objecttag"."_name") AS "taxonomy_name", "oel_tagging_taxonomy"."id", "oel_tagging_taxonomy"."name", "oel_tagging_taxonomy"."description", "oel_tagging_taxonomy"."enabled", "oel_tagging_taxonomy"."allow_multiple", "oel_tagging_taxonomy"."allow_free_text", "oel_tagging_taxonomy"."visible_to_authors", "oel_tagging_taxonomy"."export_id", "oel_tagging_taxonomy"."_taxonomy_class", "oel_tagging_tag"."id", "oel_tagging_tag"."taxonomy_id", "oel_tagging_tag"."parent_id", "oel_tagging_tag"."value", "oel_tagging_tag"."external_id", T4."id", T4."taxonomy_id", T4."parent_id", T4."value", T4."external_id", T5."id", T5."taxonomy_id", T5."parent_id", T5."value", T5."external_id" FROM "oel_tagging_objecttag" INNER JOIN "oel_tagging_taxonomy" ON ("oel_tagging_objecttag"."taxonomy_id" = "oel_tagging_taxonomy"."id") LEFT OUTER JOIN "oel_tagging_tag" ON ("oel_tagging_objecttag"."tag_id" = "oel_tagging_tag"."id") LEFT OUTER JOIN "oel_tagging_tag" T4 ON ("oel_tagging_tag"."parent_id" = T4."id") LEFT OUTER JOIN "oel_tagging_tag" T5 ON (T4."parent_id" = T5."id") LEFT OUTER JOIN "oel_tagging_tag" T6 ON (T5."parent_id" = T6."id") WHERE ("oel_tagging_objecttag"."object_id" = 'course-v1:orgA+101+test' AND NOT (NOT "oel_tagging_taxonomy"."enabled" AND "oel_tagging_taxonomy"."enabled" IS NOT NULL) AND NOT ("oel_tagging_objecttag"."taxonomy_id" IS NULL) AND NOT ("oel_tagging_objecttag"."tag_id" IS NULL AND NOT "oel_tagging_taxonomy"."allow_free_text") AND "oel_tagging_taxonomy"."enabled" AND EXISTS(SELECT 1 AS "a" FROM "content_tagging_taxonomyorg" U0 WHERE (U0."rel_type" = 'OWN' AND U0."taxonomy_id" = ("oel_tagging_objecttag"."taxonomy_id") AND (U0."org_id" IS NULL OR U0."org_id" IN (1))) LIMIT 1)) ORDER BY 8 ASC, 7 ASC
E   9. SELECT "content_tagging_taxonomyorg"."id", "content_tagging_taxonomyorg"."taxonomy_id", "content_tagging_taxonomyorg"."org_id", "content_tagging_taxonomyorg"."rel_type" FROM "content_tagging_taxonomyorg" WHERE "content_tagging_taxonomyorg"."taxonomy_id" = 5
E   10. RELEASE SAVEPOINT "s140700559607616_x2209"
E   11. SELECT "user_api_userpreference"."id", "user_api_userpreference"."user_id", "user_api_userpreference"."key", "user_api_userpreference"."value" FROM "user_api_userpreference" WHERE ("user_api_userpreference"."key" = 'pref-lang' AND "user_api_userpreference"."user_id" = 6) LIMIT 21
```
</details>

## Deadline

None

pomegranited committed:
test: fixes "list taxonomies for org" test URL

Was requesting org=$orgA instead of org=orgA, and so wasn't testing the
queries we needed to test.

pomegranited committed:
fix: filter list of taxonomies by org without fetching the Organization

openedx-webhooks commented:
Thanks for the pull request, @pomegranited! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited committed:
fix: prefetch tag_set to reduce queries

pomegranited committed:
fix: lint

rpenido commented on openedx/core/djangoapps/content_tagging/rest_api/v1/views.py:
```diff
@@ -71,7 +66,8 @@ def get_queryset(self):
 
-        return queryset.prefetch_related("taxonomyorg_set")
+        # Prefetch tag_set so we can serialize the tag counts
+        return queryset.prefetch_related("taxonomyorg_set", "tag_set")
```
We also need to prefetch the org because we match `org_short_name`. This reduced the query count by 2
 
Probably queries 13, 15, and 16 became one.
```suggestion
        return queryset.prefetch_related("taxonomyorg_set_org", "tag_set")
```

pomegranited replied:
Yep, that drops off 2 queries, thanks!

I'm considering using the RequestCache with the rules to avoid re-querying where we can, but want to use that as last resort.

pomegranited committed:
fix: prefetch taxonomyorg orgs

pomegranited committed:
temp: use openedx-learning branch from pr#157

which reduces the query count.

TO DO: once PR is merged and tagged, update the requirement here.

pomegranited committed:
fix: iterate over prefetched taxonomyorgs instead of filtering

to avoid re-querying.

Changes TaxonomyOrg.get_organizations to return the "all orgs" flag too,
which makes it useful in the rules.

pomegranited committed:
fix: Adds a check for the TaxonomyOrg.rel_type when serializing orgs

pomegranited committed:
Merge branch 'master' into jill/tagging-less-queries

pomegranited committed:
fix: oel_tagging library change reduced query count

for the taxonomy tags views.

pomegranited committed:
fix: annotate taxonomies with their tags_count

instead of prefetching all the tags.

pomegranited committed:
test: adds query count tests for non-taxonomy admins

(needs fixing)

pomegranited committed:
fix: use qs.exists() instead of len(qs) > 0

because SELECT COUNT at the database level is more efficient than
returning a bunch of data we don't need.

pomegranited committed:
fix: fetch all content library orgs at once

pomegranited committed:
fix: reduced query counts by using the user's RoleCache

instead of querying for CourseAccessRoles

Also
* removes the can_change_objecttag override method -- we only need to
  override the can_change_objecttag_taxonomy / _object+id methods to
  achieve our rules changes.
* fixes up some misleading comments, and underscores some internal
  rules method names.

pomegranited committed:
fix: adds content_tagging rules cache for Organization list

Avoids fetching and re-fetching Organizations by caching the full
Organization list for the duration of the request.

pomegranited committed:
fix: adds library orgs cache

Avoids re-fetching the libraries that a user has been granted access to
by caching the data we need from these libraries for the duration of the
request.

yusuf-musleh commented on openedx/core/djangoapps/content_tagging/rest_api/v1/serializers.py:
```diff
@@ -15,39 +15,27 @@
+    def validate(self, attrs: dict) -> dict:
+        """
+        Validate the serializer data
+       """
```
```suggestion
        """
```

yusuf-musleh commented on openedx/core/djangoapps/content_tagging/rest_api/v1/serializers.py:
```diff
@@ -86,14 +75,21 @@ class TaxonomyOrgSerializer(TaxonomySerializer):
         Return True if the taxonomy is associated with all orgs.
         """
-        return obj.taxonomyorg_set.filter(org__isnull=True).exists()
+        is_all_orgs = False
```
Unused variable: `is_all_orgs`

pomegranited replied:
Weird the linter didn't catch that!  Thanks :)

yusuf-musleh commented on openedx/core/djangoapps/content_tagging/rules.py:
```diff
@@ -268,8 +271,9 @@ def can_view_object_tag_objectid(user: UserType, object_id: str) -> bool:
 
-    object_org = Organization.objects.filter(short_name=context_key.org).first()
-    return object_org and (is_org_admin(user, [object_org]) or is_org_user(user, [object_org]))
+    assert context_key.org
```
If the `assert` for some reason fails, that would raise an `AssertionError`, is that something we want (as opposed to returning `False`) or is it handled somewhere else?

pomegranited replied:
That assert is actually to appease mypy, because `context_key.org`'s type is `Optional[str]`. But I agree, we should return False if it fails.

yusuf-musleh reviewed (approved):
:+1: This is some really great work @pomegranited ! I followed the commits while reviewing as you suggested. I just had a few minor nits and a question.

- [X] I tested this: Checked that the updated tests with less query counts pass
- [X] I read through the code
- [ ] ~I checked for accessibility issues~
- [X] Includes documentation
- [ ] ~I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.~

pomegranited commented on requirements/constraints.txt:
```diff
@@ -108,7 +108,7 @@ libsass==0.10.0
 
 # pinning this version to avoid updates while the library is being developed
-openedx-learning==0.5.1
+openedx-learning @ git+https://github.com/open-craft/openedx-learning.git@jill/tagging-less-queries
```
To do: merge & tag https://github.com/openedx/openedx-learning/pull/157, and update this dependency.

pomegranited committed:
fix: address nits in PR review

pomegranited committed:
fix: nit

bradenmacdonald commented on openedx/core/djangoapps/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -497,27 +497,37 @@ def test_create_taxonomy(self, user_attr: str, expected_status: int) -> None:
+        url = TAXONOMY_ORG_LIST_URL + f'?org={self.orgA.short_name}&enabled=true'
+        user = getattr(self, user_attr)
+        self.client.force_authenticate(user=user)
+        with self.assertNumQueries(expected_queries):
```
So the number of queries here didn't change from 16 in most cases? It that number fixed or does it increase with each additional taxonomy in the result list?

pomegranited replied:
Yes unfortunately, 16 queries was as low as I could get them.
But that number is constant -- it used to be relative to the number of TaxonomyOrgs on each returned Taxonomy.

pomegranited replied:
Hi @bradenmacdonald , did you find anything else that I need to fix in this PR?

pomegranited committed:
fix: prevent cross-org ObjectTags from being created (#633)

* fix: prevent cross-org ObjectTags from being created

A "cross-org" ObjectTag is when the object_id references an org that is not in the taxonomy's allowed list of orgs.

Similarly, we forbid creating object tags for a taxonomy with no allowed orgs listed.

bradenmacdonald commented on openedx/core/djangoapps/content_tagging/rest_api/v1/tests/test_views.py:
```diff
@@ -1759,26 +1770,43 @@ def test_get_tags(self):
-        with self.assertNumQueries(7):  # TODO Why so many queries?
+        user = getattr(self, user_attr)
+        self.client.force_authenticate(user=user)
+        with self.assertNumQueries(expected_queries):
```
Same question here - doesn't seem like the # of queries changed.

pomegranited replied:
If you expand the notes in the PR description, you can see details about what queries are still being executed.

pomegranited committed:
feat: updates openedx-learning dependency

bradenmacdonald commented on openedx/core/djangoapps/content_tagging/utils.py:
```diff
@@ -42,3 +46,57 @@ def get_context_key_from_key_string(key_str: str) -> CourseKey | LibraryLocatorV
+
+class TaggingRulesCache:
+    """
+    Caches data required for computing rules for the duration of the request.
```
This may not have been necessary if we'd have used `bridgekeeper` instead of `rules` for defining the permissions. My bad for that original mistake. It's a nice solution for now though :)

bradenmacdonald commented:
@pomegranited I went to merge this, but there's a conflict. Please fix it and I'll merge on Tuesday.

rpenido committed:
Merge branch 'master' into jill/tagging-less-queries

rpenido committed:
fix: missing import

rpenido committed:
fix: return False if user is anonymous

rpenido commented:
> @pomegranited I went to merge this, but there's a conflict. Please fix it and I'll merge on Tuesday.

Hi @bradenmacdonald! I fixed the merge conflict in case you still check it today or Monday.

openedx-webhooks commented:
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/34200&entry.1484655318=https://github.com/pomegranited&entry.752974735=2024-02-07+04:39&entry.1917517419=2024-02-16+20:15&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
